src/Orleans.TestingHost/TestStorageProviders/FaultInjectionStorageProvider.cs

```diff
@@ -1,73 +1,52 @@
 ﻿
 using System;
+using System.Dynamic;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Logging;
 using Orleans.Providers;
 using Orleans.Runtime;
 using Orleans.Storage;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
 
 namespace Orleans.TestingHost
 {
+    /// <summary>
+    /// Options for fault injection grain storage
+    /// </summary>
+    public class FaultInjectionGrainStorageOptions
+    {
+        public static TimeSpan DEFAULT_LATENCY = TimeSpan.FromMilliseconds(10);
+        /// <summary>
+        /// Latency applied on storage operation
+        /// </summary>
+        public TimeSpan Latency { get; set; } = DEFAULT_LATENCY;
+    }
     /// <summary>
     /// Fault injection decorator for storage providers.  This allows users to inject storage exceptions to test error handling scenarios.
     /// </summary>
     /// <typeparam name="TStorage"></typeparam>
-    public class FaultInjectionStorageProvider<TStorage> : IStorageProvider
-        where TStorage : IStorageProvider, new()
+    public class FaultInjectionGrainStorage : IGrainStorage, ILifecycleParticipant<ISiloLifecycle>
     {
-        private readonly TStorage realStorageProvider;
+        private readonly IGrainStorage realStorageProvider;
         private IGrainFactory grainFactory;
         private ILogger logger;
-        /// <summary>The name of this provider instance, as given to it in the config.</summary>
-        public string Name => realStorageProvider.Name;
+        private readonly FaultInjectionGrainStorageOptions options;
         
         /// <summary>
         /// Default conststructor which creates the decorated storage provider
         /// </summary>
-        public FaultInjectionStorageProvider()
+        public FaultInjectionGrainStorage(IGrainStorage realStorageProvider, string name, ILoggerFactory loggerFactory, IGrainFactory grainFactory, FaultInjectionGrainStorageOptions faultInjectionOptions)
```

Comment:
NIT: Please, please, please, when we have method/constructor that takes a lot of param, we should keep only one param per line.
It's easier to read for people with regular screen/vision, and if we have to change/add/remove one of the parameter it is way easier to read the diff

Reply:
Plus it makes merging easier.
We should auto-split lines greater than 180 chars